File: 1144-OptimizeWaterDistributionInAVillage/1144-OptimizeWaterDistributionInAVillage.cpp
// Last updated: 12/8/2025, 11:26:22 PM
class Solution {
private:
    vector<int> root;
    void init(int n) {
        root.resize(n + 1);
        for (int i = 0; i <= n; ++i)
            root[i] = i;
    }

    int find(int v) {
        return root[v] == v ? v : root[v] = find(root[v]);
    }

    bool join(int v, int u) {
        v = find(v);
        u = find(u);
        if (u == v) return false;
        root[u] = v;
        return true;
    }

public:
    int minCostToSupplyWater(int n, vector<int>& wells, vector<vector<int>>& pipes) {
        for (int i = 1; i <= n; ++i) {
            pipes.push_back({0, i, wells[i - 1]});
        }

        sort(pipes.begin(), pipes.end(), [](vector<int>& p1, vector<int>& p2) {
            return p1[2] < p2[2];
        });

        init(n);
        int cost = 0;
        for (vector<int>& p : pipes) {
            if (join(p[0], p[1]))
                cost += p[2];
        }
        return cost;
    }
};

File: 1144-OptimizeWaterDistributionInAVillage/1144-OptimizeWaterDistributionInAVillage.py
# Last updated: 12/8/2025, 11:26:24 PM
class UnionFind:
    def __init__(self, size):
        self.root = [i for i in range(size + 1)]
        self.rank = [0] * (size + 1)

    def find(self, u):
        if u != self.root[u]:
            self.root[u] = self.find(self.root[u])
        return self.root[u]

    def join(self, u, v):
        u = self.find(u)
        v = self.find(v)

        if u == v:
            return False
        else:
            if self.rank[u] > self.rank[v]:
                self.root[v] = u
            elif self.rank[u] < self.rank[v]:
                self.root[u] = v
            else:
                self.root[u] = v
                self.rank[v] += 1
            return True

class Solution:    
    def minCostToSupplyWater(self, n: int, wells: List[int], pipes: List[List[int]]) -> int:
        edges = [i for i in pipes]
        for index, cost in enumerate(wells):
            edges.append((0, index + 1, cost))

        edges.sort(key = lambda x : x[2])

        uf = UnionFind(n)
        cost = 0
        for start, end, weight in edges:
            if uf.join(start, end):
                cost += weight
        
        return cost
        
        